File: remarboy/src/se/remar/remarboy/Memory.java
package se.remar.remarboy;

import java.util.HashSet;
import java.util.Set;

// Memory map
// 0x0000-0x7FFF    Game rom (32 kB)
// 0x8000-0x9FFF    VRAM (8 kB)
// 0xA000-0xBFFF    External Expansion Working RAM (8 kB)
// 0xC000-0xDFFF    Unit Working RAM (8 kB)
// 0xE000-0xFDFF    (Use prohibited)
// 0xFE00-0xFE9F    OAM (160 bytes)
// 0xFF00-0xFF7F    Port/Mode/Control/Sound Registers
// 0xFF80-0xFFFE    Working & Stack RAM (127 bytes)
// 0xFFFF           Interrupt Enable

public class Memory {
    public int cycles;

    private Cart cart;

    private byte[] ram = new byte[0x8000];
    private byte[] metarom;

    private Set<Integer> visited = new HashSet<Integer>();

    private static final byte VISITED = 1;

    public void insertCart(String path) {
        addCart(new Cart(path));
    }

    public void addCart(Cart cart) {
        this.cart = cart;
        // TODO: Metarom will have to be more competent than this when dealing with MBCs...
        metarom = new byte[Math.max(cart.size(), 65536)];
    }

    public byte getOp(int address) {
        // TODO: Keep rom and ram separate in metarom (yes, code will be placed in ram from rom...)
        metarom[address] = VISITED;
        return getByte(address);
    }

    public byte getByte(int address) {
        if(address < 0x8000) {
            return cart.get(address);
        } else {
            return ram[address - 0x8000];
        }
    }

    public int getWord(int i) {
        return (getByte(i) & 0xff) + (getByte(i+1) & 0xff) * 0x100;
    }

    public void putByte(int address, int data) {
        if(address >= 0x8000) { // Ignore other writes for now
            ram[address - 0x8000] = (byte) (data & 0xff);
        } else {
            System.out.println(String.format("Caught write to address 0x%x", address) + ", data: " + Util.formatByte(data));
        }
    }

    public Set<Integer> getVisited() {
        visited.clear();
        for(int i = 0;i < metarom.length;i++) {
            if(metarom[i] == VISITED) {
                visited.add(i);
            }
        }
        return visited;
    }

    public byte[] getBytes(int address, int length) {
        byte[] bytes = new byte[length];
        for(int i = 0;i < length;i++) {
            bytes[i] = getByte(address + i);
        }
        return bytes;
    }
}

File: src/Memory.cpp
#include "Memory.h"
#include "Constants.h"
#include <iostream>

// Memory map
// 0x0000-0x7FFF    Game rom (32 kB)
// 0x8000-0x9FFF    VRAM (8 kB)
// 0xA000-0xBFFF    External Expansion Working RAM (8 kB)
// 0xC000-0xDFFF    Unit Working RAM (8 kB)
// 0xE000-0xFDFF    (Use prohibited)
// 0xFE00-0xFE9F    OAM (160 bytes)
// 0xFF00-0xFF7F    Port/Mode/Control/Sound Registers
// 0xFF80-0xFFFE    Working & Stack RAM (127 bytes)
// 0xFFFF           Interrupt Enable

const unsigned char VISITED = 0x01;

Memory::Memory() : metarom(0) {}

void
Memory::insertCart(std::string path) {
  cart = new Cart(path);
  cart->writeToMemory(mem, 0);
  cart->writeToMemory(&mem[BANK_SIZE], 1);

  setupMetarom();
}

unsigned char
Memory::getByte(unsigned short address) {
  return mem[address];
}

unsigned short
Memory::getWord(unsigned short address) {
  return mem[address] + (mem[address+1] << 8);
}

unsigned char
Memory::getOp(unsigned short address) {
  // TODO: Need more sophistication to deal with bank switching
  metarom[address] = VISITED;

  return mem[address];
}

void
Memory::putByte(unsigned short address, unsigned char byte) {
  if(address >= 0x8000) {
    mem[address] = byte;
  }
}

std::list<int>
Memory::getVisited() {
  std::list<int> visited;

  for(int i = 0;i < cart->getSize();i++) {
    if(metarom[i] == VISITED) {
      visited.push_back(i);
    }
  }

  return visited;
}

void
Memory::setupMetarom() {
  if(metarom != 0) {
    delete[] metarom;
  }

  metarom = new unsigned char[cart->getSize()];
  for(int i = 0;i < cart->getSize();i++) {
    metarom[i] = 0;
  }
}
